### PR TITLE
Refactor `array` and `arithm` test categories

### PR DESCRIPTION
**Pull request by canelhasmateus**
Refactor `array` and `arithm` test category.

 
Added labels and descriptions for the `array` and `arithm`, as well as completely removing echo usage from the impacted tests.

This is part of a bigger effort, see
https://github.com/nim-works/nimskull/issues/41

### CODE REVIEW

**Comment by canelhasmateus** [comment]
@zerbina

File: tests/arithm/tarithm.nim [hunk 1]
```diff
@@ -1,21 +1,30 @@
 discard """
+labels: "arithmetic bitwise conversion int range system"
+description: '''
+  . From https://github.com/nim-lang/Nim/issues/5216
```
Comment (zerbina) [inline tests/arithm/tarithm.nim:4]: Hmm, since this test file contains both regression tests and general tests, only putting the issues' summaries in the description feels off. Maybe it'd be better to use a description like:
```
description: '''
  Tests for arithmetic operations and issues related to them.
  - issue link 1
  - issue link 2
  - etc.
'''
```
and then put the issue summary in a comment next to the test case? I'm not sure.
Alternatively, the regression tests could be moved into a separate `tarithm_issues.nim` file.

I don't consider this a blocker, so it's totally fine if it stays the way it currently is.

File: tests/array/tarraycons_idx_lit_3.nim [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  line: 16
+  labels: "array constructor index"
+  description: '''
+    . Its possible to sparsely specify arrays initial values.
+      In these cases, The following indexes are based of these.
+    . Here, we create an array of size 2^32. And set
+      ( 2^32 - 1 )th value as 1,
+      ( 2^32     )th value is inferred as 2,
+      ( 2^32 + 1 )th value is inferred as 3.
+    . We obviously exceed the array size, prompting an error.
```
Comment (zerbina) [inline tests/array/tarraycons_idx_lit_3.nim]:
```suggestion
    . Here, we create an array with 3 elements. The
        1st element is at index `2^31 - 1` (explicit)
        2nd element is at index `2^31 + 0` (inferred)
        3rd element is at index `2^31 + 1` (inferred)
    . the index value of the last element (2^31 + 1) doesn't fit
      into an int32, prompting an error.
```
Slight correction: `high(int32)` is '2^31' and the array has a size of '3'.

File: tests/array/tarraycons_ptr_generic2.nim [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
   file: "tarraycons_ptr_generic2.nim"
-  line: 17
+  line: 25
+  labels: "array generic ptr ref subtyping"
+  description: '''
+    . From https://github.com/nim-lang/Nim/issues/7601
+      array construction failed generic object of ptr relation subtype with
+    . note: ref object and ref generic is ok, only ptr generic failed
+    . this bug was found during https://github.com/nim-lang/Nim/issues/4799
+      investigation
```
Comment (zerbina) [inline tests/array/tarraycons_ptr_generic2.nim]: The issue description doesn't apply for this test, even though the latter was introduced as part of the PR fixing the mentioned issue. I'm not sure what the description should be, as the test is for a compiler implementation detail (it also succeeds without the changes in the corresponding PR).

Maybe it's best to leave it empty for now?

File: tests/array/tarraycons_idx_lit_2.nim [hunk 4]
```diff
@@ -1,6 +1,10 @@
 discard """
   errormsg: "expected ordinal value for array index, got '\"string\"'"
-  line: 6
+  line: 10
   column: 10
+  labels: "array constructor index"
+  description: '''
+    . Arrays indexes should be ordinal values.
```
Comment (zerbina) [inline tests/array/tarraycons_idx_lit_2.nim]:
```suggestion
    . Array indices need to be ordinal values.
```
Small wording adjustment, since array indices being ordinal values is a requirement.

File: tests/array/tarraycons_enum.nim [hunk 5]
```diff
@@ -0,0 +1,32 @@
+discard """
+  errormsg: "invalid order in array constructor"
+  file: "tarraycons_enum.nim"
+  line: 25
+  labels: "enum array constructor"
+  description: '''
+    . When specifying an array, it is necessary to inform its static size.
+    . Since enums are statically sized , it is possible to use them to specify
```
Comment (zerbina) [inline tests/array/tarraycons_enum.nim:8]:
```suggestion
    . When creating an array, it's size needs to be known at compile-time.
    . Since enums are statically sized, it is possible to use them to specify
```
Makes it a bit easier to understand, I believe.

File: tests/array/tarraycons_idx_lit_3.nim [hunk 6]
```diff
@@ -0,0 +1,18 @@
+discard """
+  errormsg: "size of array exceeds range of index type 'range 2147483646..2147483647(int32)' by 1 elements"
+  line: 16
+  labels: "array constructor index"
+  description: '''
+    . Its possible to sparsely specify arrays initial values.
+      In these cases, The following indexes are based of these.
```
Comment (zerbina) [inline tests/array/tarraycons_idx_lit_3.nim]:
```suggestion
    . It is possible to only specify the index of the first element
      and let the compiler infer the other elements' indices.
```
A bit more precise.

**Review by zerbina — COMMENTED** [review]
Nice work :+1:. I left some wording suggestions/corrections and a thought regarding the `tarithm.nim` test.

`tarray.nim` is not reviewed yet.

File: tests/array/tarraycons_ptr_generic2.nim (hunk 3 above)
Comment (canelhasmateus) [reply tests/array/tarraycons_ptr_generic2.nim]: there are plenty of cases like this - some implementation thrown together haphazardly without much in the way of "what" or "why". Pretty hard to add meaningful descriptions to these. 
In these cases, I'm just taking whatever steps i can gather , be from the commit log or any comments.

I agree with the sentiment that this description is sub-par and dis-honest. But the alternative is to give no information. 

I don't mind leaving it empty.

File: tests/array/tarraycons_idx_lit_3.nim (hunk 2 above)
Comment (zerbina) [reply tests/array/tarraycons_idx_lit_3.nim]: Sorry, my mistake: `high(int32)` is `2^31 - 1`, so the first element is at `2^31 - 2` and not at `2^31 - 1`.

File: tests/array/tarray.nim [hunk 7]
```diff
@@ -1,116 +1,124 @@
...
+     block:
+      var arr: array[-1, int]
+      ))
 
-  x = [4, 5, 6]
-  echo repr(varArr1(x))
 
-  y = x
-  echo repr(mul(x, y))
+    proc mul(a, b: TMyarray): TMyArray =
```
Comment (zerbina) [inline tests/array/tarray.nim:135]: A `block` seems to be missing here before the proc? The tests using `mul`, `arrayGet`, and `arrayId` are for issue https://github.com/nim-lang/Nim/issues/106.

File: tests/array/tarray.nim [hunk 8]
```diff
@@ -1,116 +1,124 @@
...
+    let arr3: array[0, string] = []
+    var arr4: array[0, string] = []
 
-  var
-    x, y: TMyArray
-    o: TObj
+    doAssert(arr1.len == 0)
+    doAssert(arr2.len == 0)
+    doAssert(arr3.len == 0)
+    doAssert(arr4.len == 0)
```
Comment (zerbina) [inline tests/array/tarray.nim:123]: Did you add the `arr4` test here? I can't seem to find it in the previous version. It's no problem if it is, I just want to make sure.

File: tests/array/tarray.nim [hunk 9]
```diff
@@ -1,116 +1,124 @@
...
-  # bug #1669
-  let filesToCreate = ["tempdir/fl1.a", "tempdir/fl2.b",
-              "tempdir/tempdir2/fl3.e", "tempdir/tempdir2/tempdir3/fl4.f"]
+    proc getPos(r: TMyRecord): int =
+      result = r.x + r.y
 
-  var found: array[0..filesToCreate.high, bool]
+    doAssert sum([1, 2, 3, 4]) == 10
+    doAssert sum([]) == 0
+    doAssert getPos( (x: 5, y: 7) ) == 12
```
Comment (zerbina) [inline tests/array/tarray.nim]: Regarding the "not_sure_what_this_tests": the first two seem to test passing constant `array` values to `openArray` parameters and the other one doesn't seem to be related to arrays at all (rather to tuples). These tests date back to the very first commit, so there doesn't exists more context unfortunately.

File: tests/array/tarray.nim [hunk 10]
```diff
@@ -119,58 +127,70 @@ block tarraycons:
...
 
-  type
-    Vehicle[T] = object of RootObj
-      tire: T
-    Car[T] = object of Vehicle[T]
-    Bike[T] = object of Vehicle[T]
+    var i = 0
+    for current in fruits:
+      doAssert current.name == fruits[i][].name
+      i += 1
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
    var names: seq[string]
    for it in fruits:
      names.add(it.name)
    doAssert names == ["apple", "banana", "Fruit"]
```
The `current.name == fruits[i][].name` assertion doesn't catch if the pointers point to the wrong object somehow.

File: tests/array/tarray.nim [hunk 11]
```diff
@@ -119,58 +127,70 @@ block tarraycons:
...
+        book = Book[string](cover: "none")
+        hardBook = Hard[string](cover: "skin")
+        softBook = Soft[string](cover: "paper")
+    let books = [book, hardBook, softBook]
 
 
+    var i = 0
+    for current in books:
+      doAssert current.cover == books[i].cover
+      i += 1
```
Comment (zerbina) [inline tests/array/tarray.nim]: Same suggestion as above.

File: tests/array/tarray.nim [hunk 12]
```diff
@@ -190,70 +210,85 @@ block tarraylen:
...
+    proc fromOpenArray(x: openArray[int]) : int =
+      return x[^2]
 
-  proc mutOpenarray(x: var openArray[string]) =
-    x[^2] = "new one"
+    doAssert fromOpenArray([1, 2, 3]) == 2
 
-  useOpenarray([1, 2, 3])
+  block:
+    proc mutOpenarray(x: var openArray[string] , val : string) =
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
    proc mutOpenarray(x: var openArray[string], val: string) =
```
Tiny code-style correction.

File: tests/array/tarray.nim [hunk 13]
```diff
@@ -298,12 +333,12 @@ block troofregression:
     MySeq[T] = ref object
       data: seq[T]
 
-  proc test[T](sx: MySeq[T]) =
+  proc test[T](sx: MySeq[T]) : T =
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
  proc test[T](sx: MySeq[T]): T =
```

File: tests/array/tarray.nim [hunk 14]
```diff
@@ -536,19 +566,20 @@ block t7818:
...
 block trelaxedindextyp:
   # any integral type is allowed as index
-  proc foo(x: ptr UncheckedArray[int]; idx: uint64) = echo x[idx]
-  proc foo(x: seq[int]; idx: uint64) = echo x[idx]
-  proc foo(x: string|cstring; idx: uint64) = echo x[idx]
-  proc foo(x: openArray[int]; idx: uint64) = echo x[idx]
+  proc foo(x: ptr UncheckedArray[int]; idx: uint64) : int = x[idx]
+  proc foo(x: seq[int]; idx: uint64) : int =  x[idx]
+  proc foo(x: string|cstring; idx: uint64) : int = x[idx]
+  proc foo(x: openArray[int]; idx: uint64) : int = x[idx]
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
  proc foo(x: ptr UncheckedArray[int]; idx: uint64): int = x[idx]
  proc foo(x: seq[int]; idx: uint64): int =  x[idx]
  proc foo(x: string|cstring; idx: uint64): int = x[idx]
  proc foo(x: openArray[int]; idx: uint64): int = x[idx]
```

File: tests/array/tarray.nim [hunk 15]
```diff
@@ -562,20 +593,16 @@ block arrayLiterals:
 
 
 block t8316:
-  # https://github.com/nim-lang/Nim/issues/8316
   proc myAppend[T](a:T):string=
-    echo "a:", a
```
Comment (zerbina) [inline tests/array/tarray.nim:567]: The test also needs to make sure that `myAppend` is only called three times as was done previously via the testament output comparsion.

I'd suggest a simple `int` counter that gets incremented in `myAppend` and is asserted to be `3` after the `append2` assert.

File: tests/array/tarray.nim [hunk 16]
```diff
@@ -562,20 +593,16 @@ block arrayLiterals:
 
 
 block t8316:
```
Comment (zerbina) [inline tests/array/tarray.nim]: For consistency, this should be `block issue_8316:` I believe?

File: tests/array/tarray.nim [hunk 17]
```diff
@@ -562,20 +593,16 @@ block arrayLiterals:
...
     for a in args:
-      echo "ret:", ret
       ret.add(a)
     ret
 
-  let foo = append2("1", "2", "3")
-  echo foo
+  doAssert append2("1", "2", "3") == "123"
 
 block t12466:
```
Comment (zerbina) [inline tests/array/tarray.nim]: Rename the lables of this and following blocks to `issue_x`?

You probably just didn't get to it yet, but the issue links should likely be moved into the description as is done with the others, right?

File: tests/array/tarray.nim [hunk 18]
```diff
@@ -190,70 +210,85 @@ block tarraylen:
...
-  takeA(addr c[^1][0])
-  c[^1][1] = 5
-  echo c
+    proc deref(x: ptr int) : int = return x[]
+    doAssert deref(addr c[^1][0]) == 3
 
-  proc useOpenarray(x: openArray[int]) =
-    echo x[^2]
+  block:
+    proc fromOpenArray(x: openArray[int]) : int =
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
    proc fromOpenArray(x: openArray[int]): int =
```

File: tests/array/tarray.nim [hunk 19]
```diff
@@ -190,70 +210,85 @@ block tarraylen:
...
 
-  proc takeA(x: ptr int) = echo x[]
+    var c: seq[(int, int)] = @[(1,2), (3,4)]
+    c[^1][1] = 5
+    doAssert c == @[(1, 2), (3, 5)]
 
-  takeA(addr c[^1][0])
-  c[^1][1] = 5
-  echo c
+    proc deref(x: ptr int) : int = return x[]
```
Comment (zerbina) [inline tests/array/tarray.nim]:
```suggestion
    proc deref(x: ptr int): int = x[]
```
In general, there's no space inserted between the parenthesis and the colon. If a function only consists of a single expression, the `return` is also omitted most cases.

File: tests/array/tarray.nim [hunk 20]
```diff
@@ -119,58 +127,70 @@ block tarraycons:
...
 
-  type
-    Book[T] = ref object of RootObj
-      cover: T
-    Hard[T] = ref object of Book[T]
-    Soft[T] = ref object of Book[T]
+    var i = 0
+    for current in vehicles:
+      doAssert current.tire == vehicles[i][].tire
+      i += 1
```
Comment (zerbina) [inline tests/array/tarray.nim]: Same suggestion as above, just with a `seq[int]`.

**Review by zerbina — CHANGES_REQUESTED** [review]
Finished the review of `tarray`. Again, sorry for the delay.

There are some tests cases that don't test the same thing as they did before, plus some minor issues that need addressing, but after those are fixed, I'm good to merge.

Also, thank you for giving functions and variables more descriptive names :)

File: tests/array/tarraycons_ptr_generic2.nim (hunk 3 above)
Comment (zerbina) [reply tests/array/tarraycons_ptr_generic2.nim]: I've thought about it for a bit and I think the following description fits for the test:
```
Defensive test to make sure that the fix for https://github.com/nim-lang/Nim/issues/7601 doesn't break unrelated code.
```

File: tests/array/tarraycons_idx_lit_3.nim (hunk 2 above)
Comment (canelhasmateus) [reply tests/array/tarraycons_idx_lit_3.nim]: `high(int32)` is not `2^32 - 1`?

File: tests/array/tarraycons_ptr_generic2.nim (hunk 3 above)
Comment (canelhasmateus) [reply tests/array/tarraycons_ptr_generic2.nim]: I'll be adding this to the description.

File: tests/array/tarray.nim (hunk 8 above)
Comment (canelhasmateus) [reply tests/array/tarray.nim:123]: Yes, i added it. It seemed logical since we were already testing for const and let arrays.

File: tests/array/tarray.nim (hunk 9 above)
Comment (canelhasmateus) [reply tests/array/tarray.nim]: I'm moving the openArray to its own block and renaming this block to `leftoverTests`. I'm a fan of letting things that don't really gel together have explicitly bad names, since it eases further discussions about the validity of those. Feel free to suggest another approach.

**Comment by canelhasmateus** [comment]
Added the requested changes. Sorry for the delay, i was travelling.

File: tests/array/tarraycons_idx_lit_3.nim (hunk 2 above)
Comment (zerbina) [reply tests/array/tarraycons_idx_lit_3.nim]: No. `high(uint32)` is `2^32 - 1`, but `high(int32)` is not.

File: tests/array/tarraycons_idx_lit_3.nim [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+  line: 17
+  labels: "array constructor index"
+  description: '''
+    . It is possible to only specify the index of the first element
+      and let the compiler infer the other elements' indices.
+    . Here, we create an array with 3 elements. The
+        1st element is at index `2^31 - 1` (explicit)
+        2nd element is at index `2^31 + 0` (inferred)
+        3rd element is at index `2^31 + 1` (inferred)
+    . the index value of the last element (2^31 + 1) doesn't fit
```
Comment (zerbina) [inline tests/array/tarraycons_idx_lit_3.nim]:
```suggestion
        1st element is at index `2^31 - 2` (explicit)
        2nd element is at index `2^31 - 1` (inferred)
        3rd element is at index `2^31 + 0` (inferred)
    . the index value of the last element (2^31 + 0) doesn't fit
```

**Review by zerbina — COMMENTED** [review]
> Added the requested changes. Sorry for the delay, i was travelling.

No worries :)

After the off-by-one issue with the one issue description is fixed (sorry for my first suggestion being wrong), I'm good to merge.

File: tests/array/tarraycons_idx_lit_3.nim (hunk 21 above)
Comment (zerbina) [reply tests/array/tarraycons_idx_lit_3.nim]: Hmm, you marked the suggestion as resolved, but it wasn't applied yet. Maybe I'm missing something?

Comment (canelhasmateus) [reply tests/array/tarraycons_idx_lit_3.nim]: damned be ammends and staging areas

**Review by zerbina — APPROVED** [review]
bors r+

**Comment by canelhasmateus** [comment]
what has happened here? bors broke?

**Comment by zerbina** [comment]
The `bors` issue seems to be related to PRs that were moved out of the `draft` state. Closing and re-opening the PRs helped in the past.

bors r+
